File: riceCooker.scala
import scala.io.StdIn.readLine

class RiceCooker {
    private var isPlugged: Boolean = false
    private var powerOn: Boolean = false
    private var riceInserted: Boolean = false
    private var waterInserted: Boolean = false
    private var riceQuantity: Int = 0
    private var waterQuantity: Int = 0
    private val capacity: Int = 900

    def plug(): Unit = {
        if (isPlugged) {
            println("Rice Cooker is already plugged in.")
        } else {
            isPlugged = true
            println("Rice Cooker plugged in.")
        }
    }

    def unplug(): Unit = {
        if (!isPlugged) {
            println("The rice cooker is already unplugged.")
        } else {
            isPlugged = false
            println("The rice cooker has been unplugged.")
        }
    }

    def turnOn(): Unit = {
        if (powerOn) {
            println("The Rice Cooker is already on.")
        } else if (!isPlugged) {
            println("The rice cooker should be plugged before turning on.")
        } else if (!riceInserted) {
            println("You should insert rice.")
        } else if (!waterInserted) {
            println("You should insert water.")
        } else {
            powerOn = true
            println("Rice Cooker turned on, start cooking.")
        }
    }

    def turnOff(): Unit = {
        if (!powerOn) {
            println("The Rice Cooker is already off.")
        } else {
            powerOn = false
            println("Rice Cooker turned off.")
        }
    }

    def addRice(): Unit = {
        if (riceInserted) {
            println("Rice is already present.")
        } else {
            val quantityStr = readLine("Rice quantity: ")
            val quantity = quantityStr.toInt

            if (waterInserted) {
                if (quantity <= waterQuantity && quantity + riceQuantity <= capacity) {
                    riceQuantity = quantity
                    riceInserted = true
                    println("Rice has been added.")
                } else {
                    println("Not enough space or insufficient water quantity.")
                }
            } else {
                if (quantity <= capacity) {
                    riceQuantity = quantity
                    riceInserted = true
                    println("Rice has been added.")
                } else {
                    println("Not enough space.")
                }
            }
        }
    }

    def addWater(): Unit = {
        if (waterInserted) {
            println("Water is already present.")
        } else {
            val quantityStr = readLine("Water quantity: ")
            val quantity = quantityStr.toInt

            if (riceInserted) {
                if (quantity <= capacity - riceQuantity && quantity <= capacity) {
                    waterQuantity = quantity
                    waterInserted = true
                    println("Water has been added.")
                } else {
                    println("Not enough space or insufficient capacity for water with current rice quantity.")
                }
            } else {
                if (quantity <= capacity) {
                    waterQuantity = quantity
                    waterInserted = true
                    println("Water has been added.")
                } else {
                    println("Not enough space.")
                }
            }
        }
    }

    def getStatus(): Unit = {
        if (isPlugged) {
            println("Plugged.")
        } else {
            println("Unplugged.")
        }

        if (powerOn) {
            println("ON.")
        } else {
            println("OFF.")
        }

        if (riceInserted) {
            println(s"Rice inserted: $riceQuantity")
        }

        if (waterInserted) {
            println(s"Water inserted: $waterQuantity")
        }
    }
}

File: riceCooker.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RiceCooker = void 0;
var readlineSync = require("readline-sync");
var RiceCooker = /** @class */ (function () {
    function RiceCooker() {
        this.isPlugged = false;
        this.powerOn = false;
        this.riceInserted = false;
        this.waterInserted = false;
        this.riceQuantity = 0;
        this.waterQuantity = 0;
        this.capacity = 900;
    }
    RiceCooker.prototype.plug = function () {
        if (this.isPlugged) {
            console.log("Rice Cooker is already plugged in.");
        }
        else {
            this.isPlugged = true;
            console.log("Rice Cooker plugged in.");
        }
    };
    RiceCooker.prototype.unplug = function () {
        if (!this.isPlugged) {
            console.log("The rice cooker is already unplugged.");
        }
        else {
            this.isPlugged = false;
            console.log("The rice cooker has been unplugged.");
        }
    };
    RiceCooker.prototype.turnOn = function () {
        if (this.powerOn) {
            console.log("The Rice Cooker is already on.");
        }
        else if (!this.isPlugged) {
            console.log("The rice cooker should be plugged before turning on.");
        }
        else if (!this.riceInserted) {
            console.log("You should insert rice.");
        }
        else if (!this.waterInserted) {
            console.log("You should insert water.");
        }
        else {
            this.powerOn = true;
            console.log("Rice Cooker turned on, start cooking.");
        }
    };
    RiceCooker.prototype.turnOff = function () {
        if (!this.powerOn) {
            console.log("The Rice Cooker is already off.");
        }
        else {
            this.powerOn = false;
            console.log("Rice Cooker turned off.");
        }
    };
    RiceCooker.prototype.addRice = function () {
        if (this.riceInserted) {
            console.log("Rice is already present.");
        }
        else {
            var quantityStr = readlineSync.question("Rice quantity: ");
            var quantity = parseInt(quantityStr, 10);
            if (this.waterInserted) {
                if (quantity <= this.waterQuantity && (quantity + this.riceQuantity) <= this.capacity) {
                    this.riceQuantity = quantity;
                    this.riceInserted = true;
                    console.log("Rice has been added.");
                }
                else {
                    console.log("Not enough space or insufficient water quantity.");
                }
            }
            else {
                if (quantity <= this.capacity) {
                    this.riceQuantity = quantity;
                    this.riceInserted = true;
                    console.log("Rice has been added.");
                }
                else {
                    console.log("Not enough space.");
                }
            }
        }
    };
    RiceCooker.prototype.addWater = function () {
        if (this.waterInserted) {
            console.log("Water is already present.");
        }
        else {
            var quantityStr = readlineSync.question("Water quantity: ");
            var quantity = parseInt(quantityStr, 10);
            if (this.riceInserted) {
                if (quantity <= this.capacity - this.riceQuantity && quantity <= this.capacity) {
                    this.waterQuantity = quantity;
                    this.waterInserted = true;
                    console.log("Water has been added.");
                }
                else {
                    console.log("Not enough space or insufficient capacity for water with current rice quantity.");
                }
            }
            else {
                if (quantity <= this.capacity) {
                    this.waterQuantity = quantity;
                    this.waterInserted = true;
                    console.log("Water has been added.");
                }
                else {
                    console.log("Not enough space.");
                }
            }
        }
    };
    RiceCooker.prototype.getStatus = function () {
        if (this.isPlugged) {
            console.log("Plugged.");
        }
        if (!this.isPlugged) {
            console.log("Unplugged. ");
        }
        if (this.powerOn) {
            console.log("ON.");
        }
        if (!this.powerOn) {
            console.log("OFF");
        }
        if (this.riceInserted) {
            console.log("Rice inserted: ".concat(this.riceQuantity));
        }
        if (this.waterInserted) {
            console.log("Water inserted: ".concat(this.waterInserted));
        }
    };
    return RiceCooker;
}());
exports.RiceCooker = RiceCooker;
